Adds an esm bundle

matthewp opened the pull request:
This changes the rollup config so that it can build an esm bundle. This
excludes the canvas module which gets shimmed to a noop.

You can run with:

```
npx rollup -c rollup/es.config.j
```

I didn't add a build script or publish hook as I wasn't sure how you
felt about those, figured you'd adjust that to your liking anyways.

Adds a test that this works in a web worker. I've tested in Deno as well
and it works there too.

Cheers!

Closes #79

WebReflection commented:
Thank you! I’m off my laptop watching Europeans but I’ll have a look ASAP and publish it when I can. Appreciated ❤️

matthewp commented:
No rush at all, enjoy your weekend! Thanks for linkedom.

WebReflection commented:
Btw, I don’t deno much, is there any particular step I should do to make it available in there too? Is this MR covering it already? Thanks

WebReflection commented on .gitignore:
```diff
@@ -1,4 +1,5 @@
-node_modules/
\ No newline at end of file
+node_modules/
+es.js
```
any reason to ignore this file? it could be used even in GitHub pages to test some stuff live. Maybe `worker.js` or `web.js` would be a better name I could add in the exports dictionary too. What do you think?

matthewp replied:
No reason, I was unsure how you liked to do this. I agree that `es.js` isn't the best name. Between those I think I like `worker.js` slightly better. Let me know what you want and I'll update it.

WebReflection commented on test/assert-es.js:
```diff
@@ -0,0 +1,15 @@
+export default {
+for: Class => {
+    console.log('\x1b[1m', Class.padEnd(60, ' '), '\x1b[0m');
```
If this is going to be tested on Web, I'd rather go for:

```js
console.log('%c' + Class, 'font-weight:bold');
```

**and** I would remove the `process.exit(1)` as that would fail either ways.

matthewp replied:
Whoops, yep I'll update this, I pretty much just copied and pasted without really looking at the code.

WebReflection commented on rollup/es.config.js:
```diff
@@ -1,16 +1,46 @@
+        case 'shim:canvas': {
+          return `
+            export default {
+              createCanvas: () => {}
```
maybe we can use the callback here, so that it won't fully break when/if used.

```js
        return `
          class Canvas {
              constructor(width, height) {
                this.width = width;
                this.height = height;
              }
              getContext() { return null; }
              toDataURL() { return ''; }
            }
            export default (width, height) => new Canvas(width, height);
          `;
```

This is basically the same fallback used in `canvas.cjs`.

matthewp commented:
I've updated based on the review feedback. Went with `worker.js` and added to the export map. Let me know if you want to change that.

----

> Btw, I don’t deno much, is there any particular step I should do to make it available in there too? Is this MR covering it already? Thanks

Since Deno uses URLs for imports you can publish to npm and use in Deno using unpkg or skypack or cdnjs, those will all work without having to do anything. If you want it to feel more official you can go to https://deno.land/x and go through the steps of setting up being published there. It works based on tags on GitHub, so I think you won't have to do anything other than tag your releases (which you already do) once you have it set up.

WebReflection commented on test/assert-es.js:
```diff
@@ -0,0 +1,15 @@
+        console.error('', 'Expected:', String(expected));
+        console.error('', 'Got instead:', String(expression));
+        console.error('');
+        process.exit(1);
```
nope, I think this still need to go away from here

WebReflection replied:
Actually ... you know what? I'll drop this, as I need to go soon, no need to hang in here

matthewp replied:
Ah yeah I missed that, my bad.

WebReflection commented:
@matthewp thanks a lot! everything looks fine except the `process.exit(1)` which breaks on Web and shouldn't exist in ESM, but I don't know if deno has it even when it's not required ... that's the only thing, then I'll merge and publish 👍

WebReflection commented:
@matthewp it's up and running as `0.11` 🥳

matthewp commented:
I can confirm it's working! Thanks!

deckchairlabs commented:
@matthewp How do you get the Typescript definitions working in Deno?

matthewp commented:
@deckchairlabs I'm not sure, I don't use TypeScript.
